pkg/service/model_build_listener.go

```diff
@@ -171,3 +194,24 @@ func (t *defaultModelBuildTask) buildListenerConfig(ctx context.Context) listene
 func (t *defaultModelBuildTask) buildListenerTags(ctx context.Context) (map[string]string, error) {
 	return t.buildAdditionalResourceTags(ctx)
 }
+
+func mergeServicePortsForListener(ports []corev1.ServicePort) corev1.ServicePort {
+	port0 := ports[0]
+	mergeableProtocols := map[corev1.Protocol]bool{
+		corev1.ProtocolTCP: true,
+		corev1.ProtocolUDP: true,
+	}
+	if _, ok := mergeableProtocols[port0.Protocol]; !ok {
+		return port0
+	}
+	for _, port := range ports[1:] {
+		if _, ok := mergeableProtocols[port.Protocol]; !ok {
+			continue
+		}
+		if port.NodePort == port0.NodePort && port.Protocol != port0.Protocol {
```

Comment:
And alternative to that last would be for the logic in `(m *defaultResourceManager) registerPodEndpoints` and the `defaultEndpointResolver` APIs it calls, to validate that when given a numeric port, that for TCP_UDP two ports exist on the service, and that for non-TCP/UDP, the right one is seen (which means adding the protocol to the ServiceRef in TargetGroupBinding as right now it's not specified if the numeric port is TCP or UDP).

When given a named port for TCP_UDP, it really should be a pair (as the TCP and UDP ports will have different names) and validate that _both_ exist.

And then for TCP_UDP, exclude any resulting endpoints where the resulting port for TCP and UDP are different. Right now that won't be checked as for a numeric port, it'll take the first port name that has the right port number, irrespective of TCP or UDP) and for a named port, only one name is currently preserved through the pipeline.

Reply:
The simplest way forward of-course is to initially limit TCP_UDP support to instance-target NLBs pending further design work.

Reply:
Just to throw another small spanner in the works, [`spec.allocateLoadBalancerNodePorts`](https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation) may allow the user to make this value (`port.NodePort` and `port0.NodePort`) be 0.

That only makes sense for ip-target NLBs though, as instance-target NLBs rely on a NodePort. I'm not sure if the AWS Load Balancer Controller is going to be able to override that setting for instance-target NLBs, but a NodePort equality test should probably never two `NodePort==0` ports as equal, to avoid making an invalid situation worse.